### PR TITLE
aligator-next: new (parallel, block-sparse) Riccati solvers, CHOLMOD integration, split first- and second- order derivatives

### LINKED ISSUE #118: Refactor LQ solvers out of main solver classes

### LINKED ISSUE #119: Upgrade to C++17
After discussion, it seems it is reasonable to upgrade to C++17 (enabling CTAD, use of `<optional>` and `<filesystem>` as we see fit).

- [x] replace use of `boost::optional` with `std::optional`
- [x] introduce class template deduction guides where needed
- [x] use string_view instead of `const char*` where relevant
- [x] identify pieces of code where `if constexpr` can help simplify

@nim65s @jorisv

### LINKED ISSUE #120: Profiling
I have discussed this with @jorisv and @fabinsch a bit.

The idea is that we can use a more advanced code profiler such as [tracy](https://github.com/wolfpld/tracy) (which requires instrumenting the code) to better assess performance bottlenecks in some critical (sub-millisecond) pieces of code with high accuracy.

### LINKED ISSUE #144: Split `computeDerivatives` into first and second-order
currently the function `StageModelTpl<Scalar>::computeDerivatives` computes both first and second-order derivatives. It would be nice to have the choice of computing Hessians or [here](https://github.com/Simple-Robotics/aligator/blob/main/include/aligator/core/stage-model.hxx#L89).

### LINKED ISSUE #148: Integrate with SuiteSparse solvers
Since https://github.com/conda-forge/suitesparse-feedstock, the latest version (v7.7+) of [SuiteSparse](https://github.com/DrTimothyAldenDavis/SuiteSparse) is available. We may want to add something like CHOLMOD LDLT as a possible backend for the solvers here, using the [Eigen integration](https://eigen.tuxfamily.org/dox/group__CholmodSupport__Module.html).

### LINKED ISSUE #155: Next version of aligator
Here's a list of features coming:

- [x] #119 
- [x] #144
- [x] #148 
- [x] #120 using the [tracy](https://github.com/wolfpld/tracy) profiling tool
- [x] #118 
- [x] #156
- [x] Add block matrix/vector type for convenience #157

### LINKED ISSUE #156: Parallelized, block-sparse Riccati recursion
Implemented in our latest paper.

- [ ] depends on #118 to make changing solvers more flexible

### PR DESCRIPTION
**Pull request by ManifoldFR**
This PR:

- Splits first- and second- order derivatives (Resolves #144, @fabinsch )
- Adds [tracy](https://github.com/Simple-Robotics/tracy) as a profiling tool (Resolves #120)
- Upgrades the minimum version of C++ to 17 (Resolves #119, @ManifoldFR )
- Adds a new sublibrary called `gar` to factorize LQ subproblems for solvers (Resolves #118, @ManifoldFR )
- Adds a parallel, block-sparse factorization for the implicit/proximal Riccati algorithm (Resolves #156, @ManifoldFR )
- Integrates the CHOLMOD solver from the SuiteSparse library into `gar` (Resolves #148)
- Upgrades the minimum required version of eigenpy to 3.4
- Add a C++ example and benchmark of Talos walking by @edantec
- Major rewrite of SolverProxDDP : no longer use QFunction or ValueFunction structs, delegate Riccati recursion to `gar` solvers
- Rewrite the `Logger` class (rename from `BaseLogger`)
- Merge struct ODEData into ContinousDynamicsData, rename continous-base.hpp header to continuous-dynamics-abstract

I would like to thank the contributors to this PR: @fabinsch, @edantec, @jorisv

Resolves #155

### CODE REVIEW

**Comment by jcarpent** [comment]
Great achievement @ManifoldFR! Bravo!

**Comment by ManifoldFR** [comment]
Adding this comment here for future reference (it'll be added to the docs)

At time of writing, in the current master branch (3.4.90), the global variable which tracks whether allocations are allowed is _thread-safe_:
https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/util/Memory.h#L96

But on released versions of Eigen it is **NOT**, hence any check for mallocs when multithreading is enabled will break as on/off switches for the malloc status will create race conditions.

**Comment by ManifoldFR** [comment]
@jorisv @edantec @fabinsch you guys think we can merge this bad boy ?

File: bench/CMakeLists.txt [hunk 1]
```diff
@@ -8,16 +8,36 @@ function(create_bench exfile croc)
...
+
 create_bench("lqr.cpp" FALSE)
+create_bench("se2-car.cpp" FALSE)
 if(BUILD_CROCODDYL_COMPAT)
   create_bench("croc-talos-arm.cpp" TRUE)
   target_add_example_robot_data(bench-croc-talos-arm)
-  create_bench("multithreading.cpp" TRUE)
-  target_add_example_robot_data(bench-multithreading)
 endif()
+if(PINOCCHIO_V3)
```
Comment (LudovicDeMatteis) [inline bench/CMakeLists.txt:36]: talos-walk benchmark requires BUILD_CROCODDYL_COMPAT, should be move under the corresponding if statement

Comment (ManifoldFR) [reply bench/CMakeLists.txt:36]: Excellent catch @LudovicDeMatteis

**Review by edantec — APPROVED** [review]

**Pull request merged**